python/cornserve/services/task_registry/manager.py

```diff
@@ -0,0 +1,357 @@
+"""Runtime task and descriptor registry service.
+
+This module provides a single integration point for all services to:
+- discover task definitions and execution descriptors at runtime
+- register loaded classes into in-process registries and sys.modules
+- create and retrieve task instances by name
+
+Kubernetes specifics (custom resources, watching, etc.) are fully encapsulated
+inside this module and are not exposed through its public API.
+"""
+
+from __future__ import annotations
+
+import asyncio
+import base64
+from typing import TYPE_CHECKING, Any
+
+from kubernetes_asyncio import client, config
+from kubernetes_asyncio.watch import Watch
+
+from cornserve.constants import CRD_GROUP, CRD_VERSION, K8S_NAMESPACE
+from cornserve.logging import get_logger
+from cornserve.services.task_registry.task_class_registry import TASK_CLASS_REGISTRY
+
+if TYPE_CHECKING:
+    from cornserve.task.base import UnitTask
+
+
+logger = get_logger(__name__)
+
+
+class TaskRegistry:
+    """Service responsible for runtime registry population and access.
+
+    Public methods avoid leaking any underlying storage concepts.
+    """
+
+    def __init__(self) -> None:
+        self._api_client: client.ApiClient | None = None
+        self._custom_api: client.CustomObjectsApi | None = None
+
+    async def _load_config(self) -> None:
+        if self._api_client:
+            return
+        try:
+            config.load_incluster_config()
+            logger.info("Loaded in-cluster Kubernetes config.")
+        except config.ConfigException as e:
+            logger.error("Failed to load Kubernetes config: %s", e)
+            raise RuntimeError("Could not load Kubernetes configuration") from e
+
+        self._api_client = client.ApiClient()
+        self._custom_api = client.CustomObjectsApi(self._api_client)
+
+    async def create_task_definition(
+        self,
+        name: str,
+        task_class_name: str,
+        module_name: str,
+        source_code: str,
+        is_unit_task: bool = True,
+        namespace: str = K8S_NAMESPACE,
+    ) -> dict[str, Any]:
+        """Create a task definition from source code."""
+        await self._load_config()
+        assert self._custom_api is not None
+
+        encoded_source = base64.b64encode(source_code.encode("utf-8")).decode("utf-8")
+
+        body = {
+            "apiVersion": f"{CRD_GROUP}/{CRD_VERSION}",
+            "kind": "TaskDefinition",
+            "metadata": {"name": name, "namespace": namespace},
+            "spec": {
+                "taskClassName": task_class_name,
+                "moduleName": module_name,
+                "sourceCode": encoded_source,
+                "isUnitTask": is_unit_task,
+            },
+        }
```

Comment:
I think this (and other instances of `body` in other methods) should be made into a Pydantic model for validation, and then the Pydantic models for the CRDs are to be kept in sync with Kustomize files. Perhaps some definitions can be shared with the definitions in the Gateway models, which means we can probably move the model definitions upward in the `cornserve` module hierarchy.